### DISCUSSION TITLE
Incus/lxd and lxc in jailmaker instances

### DISCUSSION THREAD

**Discussion by Gaston-DSP** [Jip-Hop/jailmaker#75] [Q&A]
First,  thank you to everyone contributing to jailmaker and Jip-Hop in particular!
jail-maker repo is tagged with both "lxc-container" and "lxc-containers" so I was looking for some examples.
Are there any examples running Incus/lxd inside a jailmaker instance?
Is this even possible?

**Comment by Jip-Hop** [comment]
Yes there are some (very fresh) examples. I've been playing with it this weekend. Look in the develop branch of this repo inside the docs folder. 😄

**Reply by Gaston-DSP** [reply]
Very interesting and cool!
Now to the last piece of the puzzle, how do we make a delegated dataset available to the incus-host and make it available for incus.
It would be a dream to have zfs.delegate=true available to the incus-host.
I did try to --bind-ro some zfs libraries and zfs binaries to no avail. Looks like I need guidance or examples to succeed.

Many thanks again!

**Comment by Jip-Hop** [comment]
Did you bind mount /dev/zfs? In my quick testing I was able to create a zfs dataset from inside the jail (which would be the incus host) and the dataset would show up inside the TrueNAS web gui. I didn't bind mount any binaries, I just installed the zfs userspace utilities inside the jail using the package manager. I did run the jail with all capabilities added (docker_compatible=1 in the config file, I should think of a better name for that).

Let me know if you try and how it went :)

**Reply by Gaston-DSP** [reply]
```
root@incushost:~# uname -a
Linux incushost 6.1.63-production+truenas #2 SMP PREEMPT_DYNAMIC Mon Dec 18 19:34:42 UTC 2023 x86_64 GNU/Linux
```


Getting the zfsutils-linux running seems to be my issue:
I have shortened the below a bit for readability
```
root@incushost:~# apt-get install zfsutils-linux
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Package zfsutils-linux is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'zfsutils-linux' has no installation candidate
root@incushost:~# apt-get install software-properties-common -y
root@incushost:~# apt-add-repository contrib -y
root@incushost:~# apt-add-repository -L
deb http://deb.debian.org/debian bookworm contrib main
deb http://deb.debian.org/debian bookworm-updates contrib main
deb http://deb.debian.org/debian-security/ bookworm-security contrib main
root@incushost:~# apt-get install zfsutils-linux
root@incushost:~# zfs list
The ZFS modules are not loaded.
Try running '/sbin/modprobe zfs' as root to load them.
root@incushost:~# /sbin/modprobe zfs
modprobe: FATAL: Module zfs not found in directory /lib/modules/6.1.63-production+truenas
```
So it seems what I get built is a set of .ko for 6.1.0-17-amd64
If I try to force them to load I get
```
root@incushost:~# /sbin/modprobe -S 6.1.0-17-amd64 zfs
modprobe: ERROR: could not insert 'zfs': Exec format error
```
I'm sure there is something trivial I have missed?

**Reply by Jip-Hop** [reply]
You don't need to load the kernel module, it's already loaded since TrueNAS is using ZFS. Are you using the jlmkr.py file from the develop branch? It adds --bind-ro=/sys/module which is perhaps what you're missing.
